### PR TITLE
V2 STL Deserializer

### PR DESCRIPTION
**Pull request by z3dev**
This pull request updates the STL deserializer to use CSG V2 API, as well as translate to CSG V2 API. The goals of this pull request are:
- [X] Update to use CSG V2 API
- [X] Update code up to JSCAD coding standards
- [X] Add test suites to complete coverage, 93.63%

New Features:
- support for colors

Fixes jscad/io#8 
Fixes jscad/io#53 

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

**Comment by z3dev** [comment]
**BIG NOTE: ensureManaifoldness() is broken.**

Do we want this to work? If so then an additional export of retessellate() is required as part of the booleans.

File: packages/io/stl-deserializer/test/test.js [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
-// x=>console.log(
+const toArray = (polygons) => polygons.map((p) => p.vertices.map((v) => ([v[0], v[1], v[2]])))
+
 test('translate simple ascii stl to jscad code', function (t) {
   const inputPath = path.resolve(samplesPath, 'stl/testcube_ascii.stl')
   const inputFile = fs.readFileSync(inputPath, 'utf8')
 
-  const expected = `function main() { return union(
-// objects: 1
+  const expected = `function main() { return [
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/test/test.js]: This also needs the require() calls and the module.export = {main} call at the end :)

File: packages/io/stl-deserializer/test/test.js [hunk 2]
```diff
@@ -128,79 +131,95 @@ polyhedron({ points: [
...
-  const observed = deserializer.deserialize(inputFile, undefined, {output: 'csg', addMetaData: false})
-  t.deepEqual(observed.polygons.length, 10241)
-  // too many vertices for a reasonable in detail test : even a generated JSON from just the CSG vertices is almost 2 MB !
-
-  // const observedVertices = polygonsFromCsg(observed)
-  // const expectedVertices = JSON.parse(fs.readFileSync('./testExpecteds/herringbone-gear-large.json', 'utf8'))
+  const observed = deserializer.deserialize(inputFile, undefined, { output: 'csg', addMetaData: false })
+  t.is(observed.length, 1)
+  const polygons = geometry.geom3.toPolygons(observed[0])
+  t.deepEqual(polygons.length, 12744)
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/test/test.js]: how come the polygons length has changed so much . should it not be around the same as before ?

File: packages/io/stl-deserializer/test/test.js [hunk 3]
```diff
@@ -128,79 +131,95 @@ polyhedron({ points: [
...
-
-  const observedVertices = polygonsFromCsg(observed)
-  const expectedVertices = [ [ [ 1, 1, 0 ], [ 1, 0, 0 ], [ 0, 0, 0 ], [ 0, 1, 0 ] ],
-  [ [ 0, 1, 1 ], [ 0, 1, 0 ], [ 0, 0, 0 ], [ 0, 0, 1 ] ],
-  [ [ 1, 1, 1 ], [ 1, 1, 0 ], [ 0, 1, 0 ], [ 0, 1, 1 ] ],
-  [ [ 1, 1, 0 ], [ 1, 1, 1 ], [ 1, 0, 1 ], [ 1, 0, 0 ] ],
-  [ [ 1, 0, 0 ], [ 1, 0, 1 ], [ 0, 0, 1 ], [ 0, 0, 0 ] ],
-  [ [ 1, 0, 1 ], [ 1, 1, 1 ], [ 0, 1, 1 ], [ 0, 0, 1 ] ] ]
-  t.deepEqual(observedVertices, expectedVertices)
+  const observed = deserializer.deserialize(inputFile, undefined, { output: 'csg', addMetaData: false })
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/test/test.js]: this should now be 'output': 'geometry' everywhere :)

File: packages/io/io-utils/ensureManifoldness.js [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
   const transform = input => {
-    input = 'reTesselated' in input ? input.reTesselated() : input
-    input = 'fixTJunctions' in input ? input.fixTJunctions() : input // fixTJunctions also calls this.canonicalized() so no need to do it twice
+    input = 'isRetesselated' in input ? retessellate(input) : input
+    // input = 'fixTJunctions' in input ? input.fixTJunctions() : input
     return input
   }
 
-  return input.constructor !== Array ? transform(input) : input.map(transform)
+  return Array.isArray(input) ? input.map(transform) : transform(input)
```
Comment (kaosat-dev) [inline packages/io/io-utils/ensureManifoldness.js]: MUCH better !

File: packages/io/io-utils/ensureManifoldness.js [hunk 5]
```diff
@@ -1,18 +1,17 @@
...
- * @param {string} title - The title of the book.
- * @return {csg}
+ * wrapper around internal methods (in case they change) to make sure
+ * all geometry resuts in a manifold mesh
  */
 function ensureManifoldness (input) {
   const transform = input => {
-    input = 'reTesselated' in input ? input.reTesselated() : input
-    input = 'fixTJunctions' in input ? input.fixTJunctions() : input // fixTJunctions also calls this.canonicalized() so no need to do it twice
+    input = 'isRetesselated' in input ? retessellate(input) : input
```
Comment (kaosat-dev) [inline packages/io/io-utils/ensureManifoldness.js]: haha this will be 'funny' :)

File: packages/io/stl-deserializer/.eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/.eslintrc.json]: are these rules based on anything particular or 'stock' standard js ? (I do not work much with eslint so I confess my ignorance :)

File: packages/io/stl-deserializer/README.md [hunk 7]
```diff
@@ -1,13 +1,13 @@
 ## @jscad/stl-deserializer
 
-> stl deserializer for the jscad project
+> Deserializer of STL data to CSG geometries
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/README.md]: not csg anymore ? so perhaps '3d geometry for the jscad project' ? something else ?

File: packages/io/stl-deserializer/README.md [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
 
 [![npm version](https://badge.fury.io/js/%40jscad%2Fstl-deserializer.svg)](https://badge.fury.io/js/%40jscad%2Fstl-deserializer)
 [![Build Status](https://travis-ci.org/jscad/io.svg)](https://travis-ci.org/jscad/stl-deserializer)
 
 ## Overview
 
-This deserializer converts raw stl data to jscad code (that can be evaluated to CSG/CAG).
+This deserializer converts raw STL formatted data (files) to JSCAD scripts or CSG geometries.
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/README.md]: same

File: packages/io/stl-deserializer/README.md [hunk 9]
```diff
@@ -25,25 +24,22 @@ npm install @jscad/stl-deserializer
...
 ## Usage
 
-
 ```javascript
-const stlDeSerializer = require('@jscad/stl-deserializer')
+const stlDeserializer = require('@jscad/stl-deserializer')
 
 const rawData = fs.readFileSync('PATH/TO/file.stl')
-const csgData = stlDeSerializer.deserialize(rawData, undefined, {output: 'csg'})
+const geometry = stlDeserializer.deserialize(rawData, 'file.stl', {output: 'csg'})
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/README.md]: output: 'geometry' :)
do we keep the file name parameter around ? perhaps we can ditch that in V2 ? (I only vaguely recall the details tbh)

File: packages/io/stl-deserializer/README.md [hunk 10]
```diff
@@ -25,25 +24,22 @@ npm install @jscad/stl-deserializer
...
 
 ```
 
-
 ## Contribute
 
 For questions about the API, please contact the [User Group](https://jscad.xyz/forum)
 
-PRs accepted.
+Please see the README information of the OpenJSCAD.org project for how to submit bug reports or changes.
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/README.md]: much simpler , nicer !

File: packages/io/stl-deserializer/index.js [hunk 11]
```diff
@@ -66,58 +72,64 @@ function bufferToBinaryString (buffer) {
...
-  } else {
-    return buf
   }
+  return buf
 }
 
 // reliable binary detection
-function isDataBinaryRobust (data) {
-  // console.log('data is binary ?')
+const isDataBinaryRobust = (data) => {
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/index.js]: good !

File: packages/io/stl-deserializer/package.json [hunk 12]
```diff
@@ -1,11 +1,13 @@
 {
   "name": "@jscad/stl-deserializer",
   "version": "1.0.0-alpha.d0fb3056",
-  "description": "Stl deserializer for jscad project",
-  "repository": "https://github.com/jscad/io",
+  "description": "STL deserializer for JSCAD project",
+  "repository": "https://github.com/jscad/OpenJSCAD.org",
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/package.json]: yay !

File: packages/io/stl-deserializer/package.json [hunk 13]
```diff
@@ -28,15 +30,23 @@
...
   "dependencies": {
-    "@jscad/csg": "0.7.0",
-    "@jscad/io-utils": "^1.0.0-alpha.d0fb3056"
+    "@jscad/modeling": "0.0.1",
+    "@jscad/io-utils": "1.0.0-alpha.d0fb3056"
   },
   "devDependencies": {
-    "ava": "^0.25.0"
+    "ava": "^0.25.0",
+    "eslint": "^5.13.0",
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/package.json]: do we need to have these fore every single dependency ? I think some of the common deps can be moved to a higher level in Lerna ? (can be tackled later, is not essential)

File: packages/io/stl-deserializer/tests/instantiate.test.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+const samplesPath = path.dirname(require.resolve('@jscad/sample-files/package.json'))
+
+const toArray = (polygons) => polygons.map((p) => p.vertices.map((v) => ([v[0], v[1], v[2]])))
+
+test('instantiate simple ascii stl to geometry', function (t) {
+  const inputPath = path.resolve(samplesPath, 'stl/testcube_ascii.stl')
+  const inputFile = fs.readFileSync(inputPath, 'utf8')
+
+  const observed = deserializer.deserialize(inputFile, undefined, { output: 'csg', addMetaData: false })
```
Comment (kaosat-dev) [inline packages/io/stl-deserializer/tests/instantiate.test.js]: output: geometry

File: packages/io/stl-deserializer/index.js [hunk 15]
```diff
@@ -358,36 +342,34 @@ function deserializeAsciiSTL (stl, filename, version, elementFormatter) {
...
 
   return elements
 }
 
 // FIXME : just a stand in for now from scad-api, not sure if we should rely on scad-api from here ?
-function polyhedron (p) {
-  let pgs = []
-  let ref = p.triangles || p.polygons
-  let colors = p.colors || null
+const polyhedron = (p) => {
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/stl-deserializer/index.js]: I believe so. The STL is converted to a set of points and a set of faces (indexes into the points), which is exactly what polyhedron() accepts for parameters.

File: packages/io/stl-deserializer/README.md (hunk 9 above)
Comment (z3dev) [reply packages/io/stl-deserializer/README.md]: Not sure about the file name parameter. I think it only adds to the comments of the JSCAD script.

File: packages/io/stl-deserializer/tests/instantiate.test.js (hunk 14 above)
Comment (z3dev) [reply packages/io/stl-deserializer/tests/instantiate.test.js]: Done

File: packages/io/stl-deserializer/index.js [hunk 16]
```diff
@@ -247,78 +254,56 @@ function deserializeBinarySTL (stl, filename, version, elementFormatter, debug =
     vertices.push(v3)
     triangles.push(triangle)
     normals.push(no)
-    converted++
   }
 
   if (err) {
-    console.warn(`WARNING: import errors: ${err} (some triangles might be misaligned or missing)`)
+    console.log(`WARNING: import errors: ${err} (some triangles might be misaligned or missing)`)
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/stl-deserializer/index.js]: Reverted.

File: packages/io/stl-deserializer/index.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (kaosat-dev) [reply packages/io/stl-deserializer/index.js]: Ok I added a not about Polyhedron for future cleanup

File: packages/io/stl-deserializer/test/test.js (hunk 1 above)
Comment (kaosat-dev) [reply packages/io/stl-deserializer/test/test.js]: @z3dev This is one is very important, the implicit imports are gone from V2, so any piece of code using the api needs to require() what it needs explicitely

File: packages/io/stl-deserializer/.eslintrc.json (hunk 6 above)
Comment (z3dev) [reply packages/io/stl-deserializer/.eslintrc.json]: Let talk about this later when we clean up. Lint should be configured to run across all code, with the same configuration / rules.

File: packages/io/stl-deserializer/test/test.js (hunk 1 above)
Comment (z3dev) [reply packages/io/stl-deserializer/test/test.js]: Done

**Comment by z3dev** [comment]
@kaosat-dev please merge

**Pull request merged**